### PR TITLE
core: partitioned ring buffer

### PR DESCRIPTION
**Pull request by Korablev77**
Subj. First iteration of the buffer implementation.

### CODE REVIEW

**Comment by kyukhin** [comment]
Please, ask for review with green CI.

**Review by locker — CHANGES_REQUESTED** [review]
Please fix all problems found by checkpatch and remove the 'Draft' label.

File: src/lib/core/prbuf.c [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	assert((char *) record >= prbuf_linear_begin(buf));
+	return (uint32_t) ((char *) record - prbuf_linear_begin(buf));
+}
+
+/** Returns true in case buffer has at least @a size bytes until its linear end. */
+static bool
+prbuf_has_before_end(struct prbuf *buf, uint32_t size)
+{
+	assert(prbuf_linear_end(buf) >= prbuf_current_raw(buf));
+	if ((uint32_t) (prbuf_linear_end(buf) - prbuf_current_raw(buf)) >= size)
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.c:132]: Not really: `if (buf->header->offset_end + size - sizeof(struct prbuf_header) <= buf->header->size)`

File: src/lib/core/prbuf.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+void
+prbuf_iterator_create(struct prbuf *buf, struct prbuf_iterator *iter);
+
+/**
+ * Move iterator to the next entry. In case @a iter already positioned to
+ * the last entry (to be more precise - the most freshest entry with
+ * @a offset_last) - @a res is invalidated.
+ */
+int
+prbuf_iterator_next(struct prbuf_iterator *iter, struct prbuf_entry *res);
```
[the remark this answers is not in the document]
Comment (locker) [reply src/lib/core/prbuf.h:84]: The entry header cannot possibly be broken, because `prbuf_open` checks the buffer. Please remove the return value from the public API.

[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.h:84]: But `prbuf_open` checks the buffer using `iterator_next` routine..

[the remark this answers is not in the document]
Comment (locker) [reply src/lib/core/prbuf.h:84]: I know, but it isn't an excuse to add a useless return value to the public API. If you need an iterator that checks the buffer consistency, you should add an internal iterator version and implement the public version as a thin wrapper around the private version.

File: src/lib/core/prbuf.c [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+	uint32_t offset_begin;
+	/** Offset of the next byte after the last (written) record. */
+	uint32_t offset_end;
+};
+
+/**
+ * Structure representing record stored in the buffer so it has the same
+ * memory layout.
+ */
+struct prbuf_record {
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.c]: Fixed - marked as packed as well.

File: src/lib/core/prbuf.c [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+ * - The end of the buffer (in the linear sense) contains "end mark";
+ * - The minimal size of the buffer is restricted;
+ * - Iteration direction - from the oldest entry to the newest;
+ */
+
+/** A mark of unused space in the buffer: trash is located after this point. */
+static const uint32_t prbuf_end_position = (uint32_t)(-1);
+
+/** Before storing a data in the buffer we place its size (i.e. header). */
+static const size_t entry_meta_size = sizeof(uint32_t);
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.c]: Sure, fixed.

File: src/lib/core/prbuf.c [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+static const uint32_t prbuf_version = 0;
+
+/**
+ * prbuf is assumed to store all metadata in little-endian format. Beware
+ * when decoding its content. Size overhead per store operation is 4 bytes;
+ * moreover not the whole space of the buffer can be used since we do not
+ * wrap entry if it doesn't fit til the buffer's end.
+ *
+ * There are several assumptions regarding the buffer:
+ * - Buffer always contains at least one element;
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.c]: It was made in this way to avoid extra check of emptiness in prbuf_prepare().
Reworked: now at the start there's no entries; introduced prbuf_is_empty() function.

File: src/lib/core/prbuf.c [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+{
+	buf->header = (struct prbuf_header *) mem;
+	if (! prbuf_check(buf))
+		return -1;
+	return 0;
+}
+
+static struct prbuf_record *
+prbuf_skip_record(struct prbuf *buf, struct prbuf_record *current,
+		  ssize_t to_store)
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.c:228]: To avoid overflow check and extra variable def in the cycle below.

File: src/lib/core/prbuf.c [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+{
+	assert(to_store > 0);
+	assert(to_store <= buf->header->size);
+	struct prbuf_iterator iter = {
+		.buf = buf,
+		.current = current,
+	};
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.c]: Ok, reworked that part: I've introduced internal prbuf_record_next() which operates directly on prbuf_record's.

File: src/lib/core/prbuf.c [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+	assert(size > 0);
+	uint32_t alloc_size = prbuf_record_alloc_size(size);
+	if (alloc_size > (buf->header->size - sizeof(struct prbuf_header)))
+		return NULL;
+	if (prbuf_has_before_end(buf, alloc_size)) {
+		/* Head points to the byte right after the last written entry. */
+		char *head = prbuf_current_raw(buf);
+		struct prbuf_record *next = prbuf_first_record(buf);
+		/* free_space can be overflowed - it's completely OK. */
+		uint32_t free_space = (char *) next - head;
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/lib/core/prbuf.c]: Ok, fixed.

File: test/unit/prbuf.c [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	       ! prbuf_entry_is_invalid(&entry)) {
+		if (entry.size != payload_size) {
+			rc = WRONG_PAYLOAD_SIZE;
+			goto finish;
+		}
+		if (memcmp(payload, entry.ptr, payload_size) != 0) {
+			rc = WRONG_PAYLOAD_CONTENT;
+			goto finish;
+		}
+	};
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply test/unit/prbuf.c:82]: Added test case(s).

File: test/unit/prbuf.c [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+ * There are three possible configurations of test:
+ * 1. The size of buffer;
+ * 2. The size of payload;
+ * 3. The number of saves to the buffer.
+ */
+int
+main(void)
+{
+	payload_large_init();
+	test_buffer_foreach_size();
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply test/unit/prbuf.c:257]: Added all of these cases.

**Comment by locker** [comment]
Please fix all remaining issues reported by checkpatch and mark the PR as "Ready for review" (so that it isn't a draft):

image

**Comment by Korablev77** [comment]
@locker done

File: src/lib/core/prbuf.c [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+	return buf->header->end == 0;
+}
+
+/**
+ * Assuming @a record pointing to the valid buffer position, move it to the
+ * next entry. In case of current entry is broken (contains wrong size or
+ * points out of the buffer) function returns -1.
+ */
+static int
+prbuf_record_next(struct prbuf *buf, struct prbuf_record **record)
```
Comment (locker) [inline src/lib/core/prbuf.c:168]: Nit: Please rename to `prbuf_next_record`, for consistency with `prbuf_skip_record`, `prbuf_prepare_record`, etc.

File: src/lib/core/prbuf.c [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+{
+	if (prbuf_has_before_end(buf, record_size_overhead))
+		prbuf_current_record(buf)->size = prbuf_end_position;
+}
+
+/** Store entry's size. */
+static void *
+prbuf_prepare_record(struct prbuf_record *record, size_t size)
+{
+	record->size = size;
```
Comment (locker) [inline src/lib/core/prbuf.c:262]: AFAIU this must be done on commit, not on prepare.

Please add one more test case:
1. Prepare a record without committing it.
2. Iterate over all records.

The iteration shouldn't include the prepared record, nor the records purged by prepare.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/lib/core/prbuf.c (hunk 12 above)
Comment (alyapunov) [reply src/lib/core/prbuf.c:262]: I guess no: that size is used in commit. We have to store the size somewhere because `commit` doesn't take the size as an argument; and this is the perfect place to store it.

File: src/lib/core/prbuf.c [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+prbuf_linear_begin(struct prbuf *buf)
+{
+	return (char *)buf->header + sizeof(struct prbuf_header);
+}
+
+/** Returns pointer to the next byte after the last written record. */
+static char *
+prbuf_current_raw(struct prbuf *buf)
+{
+	return prbuf_linear_begin(buf) + buf->header->end;
```
Comment (alyapunov) [inline src/lib/core/prbuf.c:97]: Nit: if you measure offset from the last byte of header (in my implementation I used offsets from the beginning of header) - you can add `char data[];` to prbuf_header and and use it instead of prbuf_linear_begin.

File: src/lib/core/prbuf.c (hunk 12 above)
Comment (locker) [reply src/lib/core/prbuf.c:262]: IMO the perfect place to store the size would be `struct prbuf` (the one that isn't persisted; allocated on stack).

File: src/lib/core/prbuf.c (hunk 11 above)
Comment (Korablev77) [reply src/lib/core/prbuf.c]: Ok, renamed.

File: src/lib/core/prbuf.c (hunk 12 above)
Comment (Korablev77) [reply src/lib/core/prbuf.c:264]: Added two test cases: see test_buffer_prepared() and test_buffer_prepared_large().
Also added following fixes:
In `prbuf_skip_record()` added NULL check since in corner case we may reach the end of the buffer (see `prepared_large test()` case);
In `prbuf_prepare`:
 - correct size of allocation passed to `prbuf_skip_record` should be `alloc_size - free_space` (since space between head and first record is accounted as well);
 - for the same reason as I've patched `prbuf_skip_record()` I also have to move `end` position in case entry does not fit till the end of the buffer (`prbuf_prepare`) - to the linear begin (`buf->header->end = 0;`). Otherwise we won't be able to say that the buffer is empty (case `prepared_large()`).

File: src/lib/core/prbuf.c (hunk 13 above)
Comment (Korablev77) [reply src/lib/core/prbuf.c:97]: I know but don't really like to introduce new members to header; moreover `linear_begin()` looks quite 'natural' to me. So if you don't mind I skip your comment.

**Comment by Korablev77** [comment]
Debug build fail doesn't look related to the patch...

**Review by locker — APPROVED** [review]
